File: gds/src/core/utils/progress/tasks/TaskProgressLogger.ts
import { Log } from "@/utils/Log";
import { Concurrency } from "@/concurrency/Concurrency";
import { BatchingProgressLogger } from "../BatchingProgressLogger";
import { ProgressLogger } from "../ProgressLogger";
import { Task } from "./Task";
import { TaskVisitor } from "./TaskVisitor";
import { LeafTask } from "./LeafTask";
import { IterativeTask, IterativeTaskMode } from "./IterativeTask";

/**
 * Specialized progress logger for task-based progress tracking.
 * Extends BatchingProgressLogger with task hierarchy awareness.
 */
export class TaskProgressLogger extends BatchingProgressLogger {
  private readonly baseTask: Task;
  private readonly loggingLeafTaskVisitor: TaskVisitor;

  constructor(
    log: Log,
    baseTask: Task,
    concurrency: Concurrency,
    leafTaskVisitor?: TaskVisitor
  ) {
    super(log, baseTask, concurrency);
    this.baseTask = baseTask;
    this.loggingLeafTaskVisitor =
      leafTaskVisitor || new LoggingLeafTaskVisitor(this);
  }

  /**
   * Log the beginning of a subtask.
   */
  public logBeginSubTask(task: Task, parentTask: Task | null): void {
    const taskName = this.taskDescription(task, parentTask);

    if (parentTask === null) {
      this.logStart(taskName);
    } else {
      this.startSubTask(taskName);
    }

    this.reset(task.getProgress().volume);
  }

  /**
   * Log the successful completion of a subtask.
   */
  public logEndSubTask(task: Task, parentTask: Task | null): void {
    const taskName = this.taskDescription(task, parentTask);

    this.log100OnLeafTaskFinish(task);

    if (parentTask === null) {
      this.logFinish(taskName);
    } else {
      this.finishSubTask(taskName);
    }
  }

  /**
   * Log the failure of a subtask.
   */
  public logEndSubTaskWithFailure(task: Task, parentTask: Task | null): void {
    const taskName = this.taskDescription(task, parentTask);

    this.log100OnLeafTaskFinish(task);

    if (parentTask === null) {
      this.logFinishWithFailure(taskName);
    } else {
      this.logFinishSubtaskWithFailure(taskName);
    }
  }

  /**
   * Generate task name for bounded iterations (FIXED/DYNAMIC modes).
   */
  private boundedIterationsTaskName(
    iterativeTask: IterativeTask,
    task: Task
  ): string {
    const maxIterations = iterativeTask.getMaxIterations();
    const currentIteration = iterativeTask.getCurrentIteration() + 1;

    return `${this.taskDescription(
      task
    )} ${currentIteration} of ${maxIterations}`;
  }

  /**
   * Generate task name for unbounded iterations (OPEN mode).
   */
  private unboundedIterationsTaskName(
    iterativeTask: IterativeTask,
    task: Task
  ): string {
    const currentIteration = iterativeTask.getCurrentIteration() + 1;

    return `${this.taskDescription(task)} ${currentIteration}`;
  }

  /**
   * Generate appropriate task description based on parent context.
   */
  private taskDescription(task: Task, parentTask?: Task | null): string {
    if (parentTask instanceof IterativeTask) {
      const iterativeParentTask = parentTask as IterativeTask;
      const iterativeTaskMode = iterativeParentTask.getMode();

      switch (iterativeTaskMode) {
        case IterativeTaskMode.DYNAMIC:
        case IterativeTaskMode.FIXED:
          return this.boundedIterationsTaskName(iterativeParentTask, task);
        case IterativeTaskMode.OPEN:
          return this.unboundedIterationsTaskName(iterativeParentTask, task);
        default:
          throw new Error(`Enum value ${iterativeTaskMode} is not supported`);
      }
    } else {
      return this.taskDescriptionSimple(task);
    }
  }

  /**
   * Generate simple task description.
   */
  private taskDescriptionSimple(task: Task): string {
    return task === this.baseTask ? "" : task.getDescription();
  }

  /**
   * Log 100% completion for leaf tasks.
   */
  private log100OnLeafTaskFinish(task: Task): void {
    task.visit(this.loggingLeafTaskVisitor);
  }
}

/**
 * Visitor that logs 100% completion for leaf tasks.
 */
class LoggingLeafTaskVisitor implements TaskVisitor {
  private readonly progressLogger: ProgressLogger;

  constructor(progressLogger: ProgressLogger) {
    this.progressLogger = progressLogger;
  }

  public visitLeafTask(leafTask: LeafTask): void {
    this.progressLogger.logFinishPercentage();
  }

  // Other visit methods are no-ops for this visitor
  public visitIntermediateTask?(task: Task): void {
    // No-op - only interested in leaf tasks
  }

  public visitIterativeTask?(iterativeTask: IterativeTask): void {
    // No-op - only interested in leaf tasks
  }

  public visit?(task: Task): void {
    // No-op - only interested in leaf tasks
  }
}

File: gds/src/core/utils/progress/tasks/task_progress_logger.rs
//! TaskProgressLogger (Java GDS parity direction).
//!
//! Java’s `TaskProgressLogger` extends `BatchingProgressLogger` and adds
//! task-hierarchy-aware begin/end logging.

use crate::core::utils::progress::{BatchingProgressLogger, ProgressLogger};

use super::Task;

pub struct TaskProgressLogger {
    base_task: Task,
    logger: BatchingProgressLogger,
}

impl TaskProgressLogger {
    pub fn new(base_task: Task, concurrency: usize) -> Self {
        let volume = base_task.get_progress().volume();
        let logger =
            BatchingProgressLogger::new(base_task.description().to_string(), volume, concurrency);

        Self { base_task, logger }
    }

    pub fn log_begin_subtask(&mut self, task: &Task, parent_task: Option<&Task>) {
        let task_name = self.task_description(task, parent_task);

        if parent_task.is_none() {
            self.logger.log_start(&task_name);
        } else {
            self.logger.start_subtask(&task_name);
        }

        self.logger.reset(task.get_progress().volume() as i64);
    }

    pub fn log_end_subtask(&mut self, task: &Task, parent_task: Option<&Task>) {
        let task_name = self.task_description(task, parent_task);

        self.log_100_on_leaf_task_finish(task);

        if parent_task.is_none() {
            self.logger.log_finish(&task_name);
        } else {
            self.logger.finish_subtask(&task_name);
        }
    }

    pub fn log_end_subtask_with_failure(&mut self, task: &Task, parent_task: Option<&Task>) {
        let task_name = self.task_description(task, parent_task);

        self.log_100_on_leaf_task_finish(task);

        if parent_task.is_none() {
            self.logger.log_finish_with_failure(&task_name);
        } else {
            self.logger.log_finish_subtask_with_failure(&task_name);
        }
    }

    pub fn log_progress(&mut self, value: usize) {
        self.logger.log_progress_amount(value as i64);
    }

    pub fn reset(&mut self, volume: usize) {
        self.logger.reset(volume as i64);
    }

    pub fn log_message(&mut self, msg: &str) {
        self.logger.log_message(msg);
    }

    pub fn log_warning(&mut self, msg: &str) {
        self.logger.log_warning(msg);
    }

    pub fn log_debug(&mut self, msg: &str) {
        self.logger.log_debug(msg);
    }

    pub fn log_debug_lazy(&mut self, supplier: &dyn Fn() -> String) {
        self.logger.log_debug(&supplier());
    }

    pub fn release(&mut self) {
        self.logger.release();
    }

    fn task_description(&self, task: &Task, _parent_task: Option<&Task>) -> String {
        // Java treats the base task as having an empty name.
        if task.description() == self.base_task.description() {
            String::new()
        } else {
            task.description().to_string()
        }
    }

    fn log_100_on_leaf_task_finish(&mut self, task: &Task) {
        if task.is_leaf() {
            self.logger.log_finish_percentage();
        }
    }
}
